app/searchers/responses_searcher.rb

```diff
@@ -91,6 +91,22 @@ def apply
       fragment.presence || "'00000000-0000-0000-0000-000000000000'"
     end
 
+    save_filter_data(search)
+
     relation.where(sql)
   end
+
+  # Save specific data that can be used for search filters.
+  def save_filter_data(search)
+    search.expressions.each do |expression|
+      if expression.qualifier.assoc.include?(:forms)
+        form_names = expression.values_list
+        form_ids = form_names.map { |name| Form.find_by(name: name).try(:id) }
+          .reject(&:blank?)
+        self.form_ids += form_ids
+      else
+        self.advanced_text += " #{expression.qualifier_text}:(#{expression.values})"
+      end
+    end
+  end
```

Comment:
This is definitely moving in the right direction. We need to talk about an important edge case though—OR expressions. We should only be plucking stuff from qualifiers if they are part of top-level ANDed expressions. (I realized this is handwavy and we should talk more about it to make it less handwavy!)

Reply:
I think we also need some model specs testing this extraction stuff. Should be pretty easy to write.